### DISCUSSION TITLE
Need help restoring `.nupkg` files having zero size

### DISCUSSION THREAD

**Discussion by jpluimers** [chocolatey/choco#2765] [Q&A] [answered]
Not sure yet what exactly happened, but during a low disk space situation caused by [Patch Tuesday](https://en.wikipedia.org/wiki/Patch_Tuesday), all of my `.nupkg` files got zero size except `choco-cleaner.nupkg`.

Now Chocolatey thinks there is only that non-zero package file is installed:

```
C:\bin\bin>choco list --localonly
Chocolatey v0.11.3
choco-cleaner 0.0.8.4
1 packages installed.
```

Found out as I was uninstalling Mercurial with `choco uninstall hg` getting this error:

``` text
C:\bin\bin>choco uninstall hg
Chocolatey v0.11.3
Uninstalling the following packages:
hg

Chocolatey uninstalled 0/0 packages.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
Unable to read package from path 'hg\hg.nupkg'.
```

Then checked the files, they were all stamped the same time-stamp of the uninstall:

``` text
C:\ProgramData\chocolatey>dir /s /b *.*pkg
C:\ProgramData\chocolatey\lib\7zip\7zip.nupkg
C:\ProgramData\chocolatey\lib\7zip.install\7zip.install.nupkg
C:\ProgramData\chocolatey\lib\beyondcompare\beyondcompare.nupkg
C:\ProgramData\chocolatey\lib\choco-cleaner\choco-cleaner.nupkg
C:\ProgramData\chocolatey\lib\chocolatey-compatibility.extension\chocolatey-compatibility.extension.nupkg
C:\ProgramData\chocolatey\lib\chocolatey-core.extension\chocolatey-core.extension.nupkg
C:\ProgramData\chocolatey\lib\chocolatey-dotnetfx.extension\chocolatey-dotnetfx.extension.nupkg
C:\ProgramData\chocolatey\lib\curl\curl.nupkg
C:\ProgramData\chocolatey\lib\dotnet4.7.1\dotnet4.7.1.nupkg
C:\ProgramData\chocolatey\lib\dotnetfx\dotnetfx.nupkg
C:\ProgramData\chocolatey\lib\Everything\Everything.nupkg
C:\ProgramData\chocolatey\lib\fiddler\fiddler.nupkg
C:\ProgramData\chocolatey\lib\gimp\gimp.nupkg
C:\ProgramData\chocolatey\lib\git\git.nupkg
C:\ProgramData\chocolatey\lib\git.install\git.install.nupkg
C:\ProgramData\chocolatey\lib\hg\hg.nupkg
C:\ProgramData\chocolatey\lib\imagemagick\imagemagick.nupkg
C:\ProgramData\chocolatey\lib\imagemagick.app\imagemagick.app.nupkg
C:\ProgramData\chocolatey\lib\InkScape\InkScape.nupkg
C:\ProgramData\chocolatey\lib\KB2919355\KB2919355.nupkg
C:\ProgramData\chocolatey\lib\KB2919442\KB2919442.nupkg
C:\ProgramData\chocolatey\lib\notepadplusplus\notepadplusplus.nupkg
C:\ProgramData\chocolatey\lib\notepadplusplus.install\notepadplusplus.install.nupkg
C:\ProgramData\chocolatey\lib\paint.net\paint.net.nupkg
C:\ProgramData\chocolatey\lib\sed\sed.nupkg
C:\ProgramData\chocolatey\lib\SourceTree\SourceTree.nupkg
C:\ProgramData\chocolatey\lib\sysinternals\sysinternals.nupkg
C:\ProgramData\chocolatey\lib\vcredist2010\vcredist2010.nupkg
```

and

``` text
 Directory of C:\ProgramData\chocolatey\lib\hg\

07/14/2022  01:45 PM                 0 hg.nupkg
```

So I got two questions:

- Is there a `choco` command that re-downloads these `.nupkg` files?
- Is there an easy way to find the download locations of these `.nupkg` then manually download them?

--jeroen

**Comment by corbob** [comment]
`choco install` has a `--skip-powershell` parameter that installs the package without running the PowerShell scripts. With this in mind, you could probably reinstall everything with a PowerShell script like:

```powershell
Get-ChildItem $env:ChocolateyInstall/lib/*/*.nupkg | % {
  choco install $_.basename --skip-powershell
}
```

**Reply by jpluimers** [reply]
> `choco install` has a `--skip-powershell` parameter that installs the package without running the PowerShell scripts. With this in mind, you could probably reinstall everything with a PowerShell script like:
> 
> ```powershell
> Get-ChildItem $env:ChocolateyInstall/lib/*/*.nupkg | % {
>   choco install $_.basename --skip-powershell
> }
> ```

Thanks for responding so quickly.

I tried this (see log gist at [Restoring zero-sized `.nupkg` files: related https://github.com/chocolatey/choco/discussions/2765](https://gist.github.com/jpluimers/a469744d5a4cafa810b31b8c6e164819)) and though it kind of works, the installed package list stays the same:

``` text
C:\bin\bin>choco list --localonly
Chocolatey v0.11.3
choco-cleaner 0.0.8.4
1 packages installed.
```

Any more `choco` parameters I can try to override the condition that outputs for instance `Unable to read package from path 'hg\hg.nupkg'.`?

--jeroen

**Comment by TheCakeIsNaOH** [comment, marked as answer]
It's a known issue that Chocolatey can't handle zero size or corrupted `.nupkg` files. 
https://github.com/chocolatey/choco/issues/717

So, I'd suggest this:
1. Writing out a list of commands to run to reinstall the packages. Run the below command, and copy the output to a text file or similar.
```
Get-ChildItem $env:ChocolateyInstall/lib/*/*.nupkg | % {
    Write-Host "choco install $($_.basename) --skip-powershell"
}
```
2. Then delete all of the zero length `.nupkg` files. There is a oneliner in the linked issue.
3. Then run the commands copied previously to re-install the packages.

**Reply by jpluimers** [reply]
> It's a known issue that Chocolatey can't handle zero size or corrupted `.nupkg` files. #717
> 
> So, I'd suggest this:
> 
> 1. Writing out a list of commands to run to reinstall the packages. Run the below command, and copy the output to a text file or similar.
> 
> ```
> Get-ChildItem $env:ChocolateyInstall/lib/*/*.nupkg | % {
>     Write-Host "choco install $($_.basename) --skip-powershell"
> }
> ```
> 
> 2. Then delete all of the zero length `.nupkg` files. There is a oneliner in the linked issue.
> 3. Then run the commands copied previously to re-install the packages.

Thanks. This seems to do the job.

Found the on-liner too: <https://github.com/chocolatey/choco/issues/717#issuecomment-753613630>

``` PowerShell
ChildItem C:\ProgramData\chocolatey\lib\*\*.nu* | where Length -eq 0 | Del
```

Need to pack for a long weekend now, but will try to combine the two later.

For now this is the list of packages to "re-install":

``` text
choco install 7zip --skip-powershell
choco install 7zip.install --skip-powershell
choco install beyondcompare --skip-powershell
choco install choco-cleaner --skip-powershell
choco install chocolatey-compatibility.extension --skip-powershell
choco install chocolatey-core.extension --skip-powershell
choco install chocolatey-dotnetfx.extension --skip-powershell
choco install curl --skip-powershell
choco install dotnet4.7.1 --skip-powershell
choco install dotnetfx --skip-powershell
choco install Everything --skip-powershell
choco install fiddler --skip-powershell
choco install gimp --skip-powershell
choco install git --skip-powershell
choco install git.install --skip-powershell
choco install hg --skip-powershell
choco install imagemagick --skip-powershell
choco install imagemagick.app --skip-powershell
choco install InkScape --skip-powershell
choco install KB2919355 --skip-powershell
choco install KB2919442 --skip-powershell
choco install notepadplusplus --skip-powershell
choco install notepadplusplus.install --skip-powershell
choco install paint.net --skip-powershell
choco install sed --skip-powershell
choco install SourceTree --skip-powershell
choco install sysinternals --skip-powershell
choco install vcredist2010 --skip-powershell
```

--jeroen

**Reply by jpluimers** [reply]
Somehow the `choco install 7zip --skip-powershell` form would not cut it for any of the installs as then still Chocolatey would think the package was not installed.

When using the form `choco upgrade --force --yes 7zip` it worked splendid (but had to download and fully install, but that was OK for me).
